site/en/blog/devtools-better-angular-debugging/index.md

```diff
@@ -0,0 +1,208 @@
+
+```json/3
+{
+  ...
+  "sources": ["foo.js", "lib.js"],
+  "x_google_ignoreList": [1],
+  ...
+}
+```
+
+The new `x_google_ignoreList` field contains a single index referring to the `sources` array: 1. This specifies that the regions mapped to `lib.js` are in fact third-party code that should be automatically added to the ignore list.
+
+In a more complex example the indices 2, 4 and 5 specify that regions mapped to `lib1.ts`, `lib2.coffee` and `hmr.js` are all third-party code that should be automatically added to the ignore list.
+
+```json/3
+{
+  ...
+  "sources": ["foo.html", "bar.css", "lib1.ts", "baz.js", "lib2.coffee", "hmr.js"],
+  "x_google_ignoreList": [2, 4, 5],
+  ...
+}
+```
+
+If you’re a framework or bundler developer, make sure the source maps generated during the build process include this field in order to hook into these new capabilities in Chrome DevTools.
+
+{% Aside %}
+There may be regions in the generated code that aren’t mapped to any original code at all. We are open to exploring the possibility of such granular ignore-listing regions. Please reach out if your framework or bundler would benefit from fine-grained ignore-listing!
+{% endAside %}
+
+### `x_google_ignoreList` in Angular
+
+As of [Angular v14.1.0](https://github.com/angular/angular-cli/releases/tag/14.1.0), the contents of the `node_modules` and `webpack` folders have been marked as _“to ignore”_.
+
+This was achieved through [a change in `angular-cli`](https://github.com/angular/angular-cli/commit/b5f6d862b95afd0ec42d9b3968e963f59b1b1658) by [creating a plugin that hooks into  webpack’s `Compiler` module](https://webpack.js.org/api/compiler-hooks/)
+
+The [webpack plugin](https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/src/webpack/plugins/devtools-ignore-plugin.ts) that our engineers created hooks into the `PROCESS_ASSETS_STAGE_DEV_TOOLING` stage and populates the `x_google_ignoreList` field in the source maps for the final assets  that webpack generates and the browser loads.
+
+```js
+const map = JSON.parse(mapContent) as SourceMap;
+const ignoreList = [];
+
+for (const [index, path] of map.sources.entries()) {
+  if (path.includes('/node_modules/') || path.startsWith('webpack/')) {
+    ignoreList.push(index);
+  }
+}
+
+map[`x_google_ignoreList`] = ignoreList;
+compilation.updateAsset(name, new RawSource(JSON.stringify(map)));
+```
+
+## Linked stack traces
+
+Stack traces answer the question of _“how did I get here”_, but oftentimes this is from the machine’s perspective, and not necessarily something that matches the developer’s perspective or their mental model of the application runtime. This is especially true when some operations are scheduled to happen asynchronously later: it could still be interesting to know the “root cause” or the scheduling side of such operations, but that’s exactly something that won’t be part of a synchronous stack trace.
```

Comment:
```suggestion
Stack traces answer the question of _“how did I get here”_, but often this is from the machine’s perspective and not necessarily something that matches the developer’s perspective or their mental model of the application runtime. This is especially true when some operations are scheduled to happen asynchronously later. It could still be interesting to know the “root cause” or the scheduling side of such operations but that’s exactly something that won’t be part of a synchronous stack trace.
```